[Clang] VectorExprEvaluator::VisitCallExpr / InterpretBuiltin - Allow AVX512 VPSHUFBITQMB intrinsics to be used in constexpr

chaitanyav opened the pull request:
Resolves: #161337

chaitanyav committed:
[Clang] VectorExprEvaluator::VisitCallExpr / InterpretBuiltin - Allow AVX512 VPSHUFBITQMB intrinsics to be used in constexpr

Resolves: #161337

RKSimon commented on clang/lib/AST/ByteCode/InterpBuiltin.cpp:
```diff
@@ -3391,6 +3391,72 @@ static bool interp__builtin_ia32_shuffle_generic(
+  APSInt RetMask(llvm::APInt(RetWidth, 0), /*isUnsigned=*/true);
+
+  for (unsigned QWordId = 0; QWordId != NumQWords; ++QWordId) {
+
```
(style) remove empty line

RKSimon reviewed (changes requested):
Mainly a few style comments (applicable to both InterpBuiltin and ExprConstant implementations).

chaitanyav committed:
Address review feedback

- Rename ByteInQWord to ByteIdx for loop counter within qword
- Add SelIdx for absolute byte index used in mask selection
- Use APInt::insertBits() instead of manual shift/OR for building SourceQWord
- Remove unnecessary const from local variables
- Move constexpr after __attribute__ in header macros
- Minor formatting cleanup

RKSimon reviewed (approved):
LGTM - cheers

RKSimon committed:
Merge branch 'main' into issue_161337
